pkg/odo/cli/component/create_methods.go

```diff
@@ -0,0 +1,393 @@
+	registryUtil "github.com/openshift/odo/pkg/odo/cli/registry/util"
+	"github.com/openshift/odo/pkg/preference"
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/zalando/go-keyring"
+	"k8s.io/klog"
+)
+
+type CreateMethod interface {
+	// CheckConflicts checks for conflicts specific to a create method
+	CheckConflicts(co *CreateOptions, args []string) error
+	// FetchDevfileAndCreateComponent fetches devfile from registry, or a remote location, or a local file system, and creates a component
+	// This method also updates the CreateOptions structure with co.devfileMetadata
+	FetchDevfileAndCreateComponent(co *CreateOptions, cmd *cobra.Command, args []string) error
+	// Rollback cleans the component context of any files that were created by odo
+	Rollback(componentContext string)
+}
+
+// InteractiveCreateMethod is used while creating a component interactively
+type InteractiveCreateMethod struct{}
+
+func (icm InteractiveCreateMethod) CheckConflicts(co *CreateOptions, args []string) error {
+	return nil
+}
+
+func (icm InteractiveCreateMethod) FetchDevfileAndCreateComponent(co *CreateOptions, cmd *cobra.Command, args []string) error {
+	catalogDevfileList, err := validateAndFetchRegistry(co.devfileMetadata.devfileRegistry.Name)
+	if err != nil {
+		return err
+	}
+
+	//SET METADATA
+	// Component type: We provide devfile component list to let user choose
+	componentType := ui.SelectDevfileComponentType(catalogDevfileList.Items)
+
+	// Component name: User needs to specify the component name, by default it is component type that user chooses
+	componentName := ui.EnterDevfileComponentName(componentType)
+
+	// Component namespace: User needs to specify component namespace, by default it is the current active namespace
+	var componentNamespace string
+	if cmd.Flags().Changed("project") {
+		componentNamespace, err = cmd.Flags().GetString("project")
+		if err != nil {
+			return err
+		}
+	} else {
+		client, e := genericclioptions.Client()
+		// if the user is logged in or if we have cluster information, display the default project
+		if e == nil {
+			componentNamespace = ui.EnterDevfileComponentProject(client.GetCurrentProjectName())
+		}
+	}
+
+	co.devfileMetadata.componentType = componentType
+	co.devfileName = componentType
+	co.devfileMetadata.componentName = componentName
+	co.devfileMetadata.componentNamespace = componentNamespace
+
+	co.devfileMetadata.devfileLink, co.devfileMetadata.devfileRegistry, err = findDevfileFromRegistry(catalogDevfileList, co.devfileMetadata.devfileRegistry.Name, co.devfileMetadata.componentType)
+	if err != nil {
+		return err
+	}
+	return fetchDevfileFromRegistry(co.devfileMetadata.devfileRegistry, co.devfileMetadata.devfileLink, co.DevfilePath, co.devfileMetadata.componentType, co.componentContext)
+}
+
+func (icm InteractiveCreateMethod) Rollback(componentContext string) {
+	os.RemoveAll(componentContext)
```

Comment:
Are you sure we want to remove all the files of the devfile context? The user could have source files in this place

Reply:
Philippe, what was the library you mentioned which could check diff in the directory. I am thinking to use it to decide what files should be deleted.

Reply:
I talked about this lib recently, but I'm not sure it is relevant: https://github.com/monochromegane/go-gitignore

But don't you think we just need to delete the devfile.yaml file here? The interface is about fetching this devfile.yaml only, I think it should only revert the creation of this file.

Reply:
On a second thought, I am not sure if deleting the starter project files is a good idea, we don't do it now. We can just delete the devfile and env files for now.